### PR TITLE
Add ThinLTO/fix non-local analysis stack safety alias support

### PR DESCRIPTION
**Pull request by vlad902**
Implementing alias support with ThinLTO increased the number of allocas (by count) marked stack-safe from 45% to 59% for chromium.

### CODE REVIEW

File: llvm/lib/Analysis/StackSafetyAnalysis.cpp [hunk 1]
```diff
@@ -233,6 +237,41 @@ SSFunctionSummary::SSFunctionSummary(FunctionSummary &FS) {
...
+SSFunctionSummary::SSFunctionSummary(AliasSummary &AS) {
+  assert(isa<FunctionSummary>(AS.getAliasee()));
+
+  this->GV = nullptr;
+  this->GVS = &AS;
+  this->DSOLocal = AS.isDSOLocal();
+  // Non-prevailing definitions are not marked live.
+  this->Interposable = !AS.isLive();
+
+  FunctionSummary &DestFn = cast<FunctionSummary>(AS.getAliasee());
```
Comment (eugenis) [inline llvm/lib/Analysis/StackSafetyAnalysis.cpp:270]: Do you need to stripPointerCasts from aliasee?
Please test alias-of-alias, where both aliasees have bitcasts.

Comment (vlad902) [reply llvm/lib/Analysis/StackSafetyAnalysis.cpp:270]: I added a testcase with a bitcasted alias and added a stripPointerCasts() to the cast for the Alias::getAliasee() in the other constructor (in this constructor FunctionSummary::getAliasee() always returns a GlobalValueSummary)

Not sure what you mean by alias-of-alias?

**Comment by eugenis** [comment]
Is it possible for an alias to point to another alias?
I think this is already supported in the current patch correctly, but a
test would not hurt.

**Comment by vlad902** [comment]
I didn't think it was allowed but upon re-reading https://llvm.org/docs/LangRef.html#aliases it seems like it actually is. And it actually shouldn't work because we verify that the AliasSummary aliasee is a FunctionSummary, but it does work right now. It looks like the ModuleSummaryAnalysis does not accurately represent aliases. I'll look to fix this and upstream it if it doesn't break anything, I'll add some additional tests here now reflecting the current ThinLTO brokenness.

**Pull request merged**